src/ocean/util/container/btree/BTree.d

```diff
@@ -0,0 +1,507 @@
+
+    - Every node may contain at most 2*degree - 1 elements. Therefore, an internal
+      node may have at most 2*degree children.
+
+    Other than insert and delete operations, inorder traversal method is added,
+    which traverses the tree in inorder order (left subtree is visited first,
+    then the element separating these subtrees, then the right subtree, then
+    the next element, etc.). This provides sequential access to the tree.
+    There's a recursive and range-based implementation of the inorder
+    traversal.
+
+    CPU complexity of the operations (n is number of the elements, t is degree,
+    h is the tree height):
+
+        - Searching: O(th) = O(t * log_t(n))
+        - Inserting: O(th) = O(t * log_t(n))
+        - Deleting:  O(th) = O(t * log_t(n))
+
+    Complexity of the memory access (to read/write the node from/in memory):
+
+        - Searching O(h) = O(log_t(n))
+        - Inserting O(h) = O(log_t(n))
+        - Deleting: O(h) = O(log_t(n))
+
+    References:
+        - https://en.wikipedia.org/wiki/B-tree
+        - Thomas H. Cormen et al. "Introduction to Algorithms, 3rd edition"
+
+    Params:
+        TreeElementType = type of the element to store.
+        tree_degree = degree of the tree (refer to the documentation)
+
+******************************************************************************/
+
+struct BTree(TreeElementType, int tree_degree)
+{
+    static assert (tree_degree > 0);
+
+    /**************************************************************************
+
+        Constructor.
+
+    ***************************************************************************/
+
+    package void init (IMemManager allocator = mallocMemManager)
+    {
+        this.impl.init(allocator);
+    }
+
+    /**************************************************************************
+
+        Inserts the element in the tree.
+
+        Params:
+            el = element to insert.
+
+        Returns:
+            true if the element did not exist and it was inserted, false otherwise
+
+    ***************************************************************************/
+
+    public bool insert (ElementType el)
+    {
+        return this.impl.insert(el);
+    }
+
+    /******************************************************************************
+
+        Deletes the element from the BTree.
+
+        Params:
+            element = element to delete.
+
+        Returns:
+            true if the element was found and deleted, false otherwise.
+
+     ******************************************************************************/
+
+    public bool remove (ref ElementType element)
+    {
+        return this.impl.remove(element);
+    }
+
+    /******************************************************************************
+
+        Finds the highest value in the tree.
+
+        Returns:
+            largest element in the tree, or ElementType.init if the tree is empty.
```

Comment:
(there's one method, I think `search` where I return the indicator if the element was found. I think this is better way).